Userland/Libraries/LibSQL/TreeNode.cpp

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2021, Jan de Visser <jan@de-visser.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Format.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/RefPtr.h>
+#include <AK/StringBuilder.h>
+
+#include <LibSQL/BTree.h>
+#include <LibSQL/Key.h>
+#include <LibSQL/Serialize.h>
+
+namespace SQL {
+
+/* ----------------------------------------------------------------------- */
+
+TreeNode::TreeNode(BTree &tree, TreeNode const* up, TreeNode *left, u32 pointer)
+    : m_tree(tree)
+    , m_up(up)
+    , m_pointer(pointer)
+    , m_entries()
+    , m_down_pointers()
+    , m_down_nodes()
+{
+    if (left) {
+        m_down_pointers.append(left->pointer());
+        m_down_nodes.append(left);
+    }
+    if (!m_pointer) {
+        m_pointer = m_tree->new_record_pointer();
+    }
+}
+
+TreeNode::TreeNode(BTree &tree, TreeNode const* up, u32 pointer, ByteBuffer &buffer, size_t &at_offset)
+    : m_tree(tree)
+    , m_up(up)
+    , m_pointer(pointer)
+    , m_entries()
+    , m_down_pointers()
+    , m_down_nodes()
+{
+    u32 nodes;
+    at_offset = deserialize_from(buffer, at_offset, nodes);
+    for (u32 i = 0; i < nodes; i++)
+    {
+        u32 left;
+        at_offset = deserialize_from(buffer, at_offset, left);
+        m_entries.append(Key(m_tree->index_def(), buffer, at_offset));
+        if (left) {
+            m_down_pointers.append(left);
+            m_down_nodes.append(nullptr);
+            VERIFY(size() == m_down_pointers.size());
+        }
+    }
+    u32 right;
+    deserialize_from(buffer, at_offset, right);
+    if (right) {
+        m_down_pointers.append(right);
+        m_down_nodes.append(nullptr);
+        VERIFY((size()  + 1) == m_down_pointers.size());
+    }
+}
+
+bool TreeNode::insert(Key const& key, TreeNode *right)
+{
+    dbgln("[{}] insert({}, {})",
+        m_pointer, (String) key, ((right) ? right->m_pointer : -1));
+    if (!is_leaf()) {
+        dbgln("Node is not a leaf. Find leaf to insert into");
+        return node_for(key)->insert(key);
+    }
+    VERIFY(is_leaf());
+    if (!m_tree->duplicates_allowed()) {
+        for (auto& entry : m_entries) {
+            if (key == entry) {
+                dbgln("BTree does not allow duplicates and key {} already exists");
```

Comment:
how did this compile, its missing an argument to fill in the fromat

Reply:
Not as far as I know, would need to ask on discord myself...

Reply:
Is your host compiler clang?
The format checks aren't enabled on clang

Reply:
gcc 10.2